### ISSUE TITLE
Entitlements from app.xml in 575

### ISSUE THREAD

**Issue by chazhenry** [airsdk/Adobe-Runtime-Support#1075]
Just upgraded to 33.1.1.575. Using IntelliJ IDE. Glad to see the InfoAdditions from app.xml making their way to the Info.plist. Still, I don't see any sign of the Entitlements getting across and thus the compiled version asks or grants no permissions. Is it me? Here is my macOS block from app.xml

`<macOS>
        <InfoAdditions>
        <![CDATA[
        <key>NSCameraUsageDescription</key>
        <string>Camera access is needed to present the coach's image on the telestrator window.</string>
        <key>NSMicrophoneUsageDescription</key>
        <string>Microphone access allows the capture of a video session.</string>
        ]]>
        </InfoAdditions>
        <Entitlements>
        <![CDATA[
        <key>com.apple.security.cs.allow-jit</key>
        <true/>
        <key>com.apple.security.cs.allow-unsigned-executable-memory</key>
        <true/>
        <key>com.apple.security.cs.disable-library-validation</key>
        <true/>
        <key>com.apple.security.device.audio-input</key>
        <true/>
        <key>com.apple.security.device.camera</key>
        <true/>
        ]]>
        </Entitlements>
    </macOS>`

**Comment by ajwfrost** [comment]
Just to confirm: the entitlements are added if you're creating an App bundle and are providing native signing options for this. So to see them, you'd need to do something like:
```
adt -package [air signing options] -target bundle -storetype KeychainStore -alias "DevID or AppStore cert name" MyApplication app.xml main.swf etc
```
and then check using
```
codesign -d --entitlements - MyApplication.app
```

thanks

**Comment by chazhenry** [comment]
@ajwfrost Sorry to be dense on this. I have a long time Windows app - with Harman fixing the mic/camera issues (thanks!), I'm trying to cross compile to Mac, but know little about MacOS.

I'm using IntelliJ and compile via Build > Package Air Application > Captive Runtime Bundle. That creates an app that runs on the Mac except for permission bound resources. In the app Contents I found Info.plist. I don't use adt and don't seem to have it installed. Are you suggesting I compile with adt? I'm finding little help on adt - what to feed the command above (ie target, storetype, etc.). What little I find is from years ago. Perhaps I should step back. What should I be using to compile a well functioning Air/Harman Windows app to the Mac?

Is there an example shell script that automates the entire pipeline? I'm not sure where the above fits into the compile I do in the IDE.

**Comment by ajwfrost** [comment]
@chazhenry  actually it sounds like you're doing everything right! I Didn't realise you were using IntelliJ: if you do the "Captive Runtime Bundle" there then under the hood, it will be calling adt with the "bundle" target, so that should all be good.

The issue would be that when they call the bundle creation, it's not doing the final signing process for the resulting .app bundle. This is one of the challenges we face when adding options/features to the ADT command-line; the IDEs like IntelliJ then don't know about them and don't support them for a little while..

The simplest (?) option might be to sign it yourself.. you'd need to use the codesign tool and create the entitlements file (you can get the existing entitlements via that command I posted earlier, and then just add in the extra lines that you need from the app descriptor file). Then to re-sign, if the entitlements are in a file "ent.plist":
```
codesign --entitlements ent.plist -s 'certificate name' -f -o runtime --strict --timestamp --deep -v MyBundle.app
```

There's some more information here including a link to a re-sign script from Distriqt which may help:
https://community.adobe.com/t5/air/error-itms-90075-submitting-to-itunes-with-fb-air-28-using-dynamic-frameworks/td-p/9703176
or the apple documentation which may be overly complicated:
https://developer.apple.com/library/archive/documentation/Security/Conceptual/CodeSigningGuide/Procedures/Procedures.html

**Comment by chazhenry** [comment]
@ajwfrost I get 'cannot read entitlement data' when using your exact entitlement plist (from previous post) and running the codesign command. From googling, it seems the plist file is binary so I must use Xcode to generate it. True? Also, no way to alter the scripts that IntelliJ uses to compile/bundle? Beaten down by this.

**Comment by ajwfrost** [comment]
Hi @chazhenry  - it should be possible to use a plain-text plist file, although I have to confess I don't tend to give them a "plist" extension - ours are "toolname.entitlements" - so I don't know whether it's just looking at the file extension and then deciding it should be a binary format..

I can check with a colleague re IntelliJ scripts, but can I ask what your certificate is that you're using for the signing options within IntelliJ? Is that a certificate you've generated via the Apple developer portal that is in your "My Certificates" view in the keychain? 

As a separate option: if you edit the AIR SDK "lib/adt.cfg" file and uncomment the line `DebugOut=true`, you can run the IntelliJ packager and then check for a file `~/adt.log` - this should then give you the command-line options that IntelliJ uses for building with ADT. If you can copy/paste that into a terminal window and get it to reproduce, then you can just add the extra native codesigning parameters after the 'bundle' parameter...

thanks

**Comment by chazhenry** [comment]
@ajwfrost no go on the entitlements file with no plist extension. same error message. search tells me that its not binary, but so particular that xcode needs to author it. no real help message to help find what's misaligned - just a general error message.  just installed xcode (20gb to edit an XML file!). 

yes. generated cert in dev portal. 

you talked about IntelliJ being behind. what IDE should a pro dev use?

thanks for the ADT tip. that might help if i can generate a plist file it likes. i really despise apple at this point. also, don't see harman as a cross platform tool with all of this duct tape and secret decoder ring BS. maybe some day. still, appreciate your help.

**Comment by ajwfrost** [comment]
Yes I'm less than impressed by some of Apple's error messages, particularly around code signing issues! and I agree on your point re the fact that this is more complex than it should be; one of our weaknesses is that we don't manage an IDE but just provide the command-line tools within the SDK, which other companies then have to integrate into their IDEs. IntelliJ is a good option but we'll always have this problem unless we keep our interfaces completely static..

.. so in this instance, I am wondering whether we could make the assumption since your AIR signing certificate is an Apple one, that we can sign the bundle with it as well. This would cut out the second step and mean that IntelliJ's current mechanism would be good enough. For folk (like our test cases!) where the AIR signing is done using a self-signed certificate, the bundle would still need to be signed separately, but this is another area where we may be able to wrap it up into ADT (so, not supported by the IDEs as it would be a new thing, but a fairly simple command-line request such as `adt -sign signing_options -target bundle MyApp.app`)

thanks

**Comment by chazhenry** [comment]
@ajwfrost This should take minutes and is taking months. I can't codesign the final app because its already codesigned (by IntelliJ), but it's a) missing the entitlements and b) is 'not bound'. I have no way of knowing if that's on the Harman side or if it's IntelliJ. I do know that the <InfoAdditions><key>NSCameraUsageDescription</key> omission was on the Harman side. Are the entitlements as well? Would the IntelliJ process work if the entitlements were correctly pulled out of the app.xml file? Why is the current app unbound?

I don't see any way to edit the IntelliJ 'package Air App' script or to unsign the compiled .app so I can re-codesign as suggested above. Soul crushing experience.

**Comment by ajwfrost** [comment]
Hi - it should always be possible to re-sign and replace an existing signature, the `codesign` tool does this if you have the `-f` option.

So I would say, if you can download [SDK version 33.1.1.620](https://airsdk.harman.com/download/33.1.1.620), you can then take the .app folder that IntelliJ has created and run this:
```
adt -sign -storetype KeychainStore -alias 'Developer ID Application' -target bundle yourapp.app signedapp.app
```

Then this should replace the code signatures and pull in the entitlements that are listed in the app descriptor file...

thanks

**Comment by chazhenry** [comment]
@ajwfrost Eureka! My main app bundle runs! I'm able to codesign with -f and add the entitlements. I have to believe there's an omission in the Harman code as IntelliJ correctly does a codesign and the <InfoAdditions> from the app.xml file are used. It's simply not using the <Entitlements> block (both are in <macOS>. As we saw earlier, it was a Harman addition that fixed the InfoAdditions. Has to be the same for <entitlements> to eliminate the need to codesign twice.

Not withstanding, I'm now stuck at a new place on Mac (it's never-ending). I run a nativeProcess in which I shell "/usr/sbin/screencapture" to capture the screen and mic. It works fine if I run from a terminal, but not if I run the main app. I'm sure it's also an entitlement, but I can't find one that works. I've added com.apple.security.temporary-exception.shared-preference.read-write and also key=com.apple.security.temporary-exception.mach-register.global-name and value=com.apple.screencapture.interactive. No dice. The program simply ignores the nativeProcess call. 

Seems Apple is going to have a line of code that lets every other line of code run. Mother may I run this code?

**Comment by ajwfrost** [comment]
@chazhenry great to hear ... yes I think we may be able to adapt what we've got, to make it work better (and single-step) so that its' easier from Intellij etc.

For the idea of launching a native process to run /usr/sbin/screencapture, we did find some other similar restrictions in the latest MacOS builds where we couldn't find a way to trigger things in certain folders... I'll check with the rest of the team but I'm not sure whether it's now possible for one application to just launch another, command-line utility... (If anyone's able to get it working within a normal/native macOS app, we should be able to work it into AIR too, which is now often how we're trying to work out some of these challenges!)

thanks

**Comment by chazhenry** [comment]
@ajwfrost I've changed the output file to the default directory of the app and now it tries to run /usr/sbin/screencapture so it may be a sandbox issue. Problem (now) is that it asks repeatedly for screen capture permission - even after its given in SystemPrefs - Security - Screen Recording. Is there doc for the new PermissionManager? What permissions are available? I could really use some sample code.

**Comment by chazhenry** [comment]
Got it. For anyone on the same path. One more entitlement for the --entitlements.plist file: 
`com.apple.security.screencapture`
added that to the follow up codesign and it allows the command line screencapture to run as a NativeProcess. It asks the first time to relax security, but with that entitlement in place, it now holds on to the setting. Hope that helps someone.

**Comment by ajwfrost** [comment]
@chazhenry  well found! Glad you got that working -> for the 'permission manager' area, the macOS APIs that we're using is just for audio and video capture:
https://developer.apple.com/documentation/avfoundation/cameras_and_media_capture/requesting_authorization_for_media_capture_on_macos?language=objc

There's nothing there to request screen capture, although we found this link - which doesn't seem particularly helpful!
https://developer.apple.com/documentation/coregraphics/3656524-cgrequestscreencaptureaccess?language=objc
plus this:
https://developer.apple.com/documentation/avfoundation/avcapturescreeninput?language=objc
but, nothing on permissions or entitlements!

So - thank you for working it out :-)

**Comment by chazhenry** [comment]
@ajwfrost  IMHO Apple has gotten the OS way in front of the dev tools. There's virtually no doc on the entitlement that I used. I just assumed it had to be there given the other `com.apple.security.whatever` assertions that I found. It further seems that OS traps and asks for permissions that are not explicitly requested since it correctly asks once if I want to allow screen recording when I make my process call. The loop I was stuck in is that without the entitlement, it asked every time. On the Harman side, I believe that it would help to pull the entitlements from the app.xml so we could centralize their maintenance. I don't know if it would work to simply put them in the Info.plist, but clearly IntelliJ is signing the bundle once. It might just work. Even if it doesn't, I'd suggest that when extracted from app.xml that it builds an Entitlements.plist file. It would still require the extra codesign, but would focus the project structure around the xml file. One of my issues was that codesign didn't like my Entitlements.plist file until I wrote it with Xcode. Classic Apple.

All told, my biggest takeaway is that the support has been excellent. This was an experiment in whether Air can be a cross platform tool for me and the answer is yes. On the software alone, it would have failed, but with your timely and consistent help and suggestions I was able to slog through what were mostly Apple issues. Thanks for the support.

**Comment by ajwfrost** [comment]
@chazhenry yes I sometimes find the documentation a little lacking.. and some inconsistencies .. makes life fun!

One thing I'm puzzled by is that you're saying IntelliJ is signing the bundle, but I'm curious how it's signing it or what with. If you take that initially-generated .app file, and run
```
codesign -d -vvvv generated.app
```
on it, can you please confirm whether the "Identifier" line shows this as your app or as an Adobe ID, and also whether you get "Authority" lines that have "Developer ID Application" in them?

What I'm wondering, is whether actually IntelliJ isn't doing anything itself, but the binary is being left with a code signature from the original template file used to create the .app binary. So this would mean we are just taking the "signing details" that you enter in IntelliJ and using them to generate the `generated.app/Contents/Resources/META-INF/signatures.xml` file...

I've just been trying to sign with a Developer ID Application certificate when using this for the initial signing options, but it's failing... if we could do this though, it might be best -> but currently it's not supported for creating bundles. The only way to sign a bundle with a Developer ID Application certificate is with the "native" signing options i.e. adding them after the `-target bundle` part of the command (which would imply a change in IntelliJ etc is required) or it would need the post-processing using that extra `-sign` call per the above.

However -> was also thinking we could add this option into the application descriptor file so if you were to include a section under
```
<macOS>
  <signingCertificate>Developer ID Application</signingCertificate>
 ...
```
or similar, then we could work out during bundle creation that we should do the post-processing. We're always on a mac at this point hence can always use the native `codesign` tool and the provided certificate name/fingerprint to do this.

thanks

**Comment by Mxlt** [comment]
Hi! I have tried following this post because I am getting very similar issues. I am also following this documentation https://airsdk.dev/docs/building/application-descriptor-files/elements/macOS but I am not sure if this is working correctly. When building it doesn't complain. But after building, I check the entitlements and the entitlements I've added aren't there at all. 

I am also using IntelliJ. I've also tried compiling without intelliJ and using the terminal, but it has the same outcome. 

By using this though `adt -sign -storetype KeychainStore -alias 'Developer ID Application' -target bundle yourapp.app signedapp.app` I was able to see some changes. The entitlements are after that correct. But for some reason the app changes and I stop being able to inspect it.

Are we sure that the documentation is correct? that is how entitlements are added?

**Comment by ajwfrost** [comment]
@Mxlt just checking what you're actually doing here: are you building using IntelliJ to create a "bundle" i.e. `youapp.app` folder? Do you know what signing options are configured for this? If you want to do a single-shot command to sign everything both from an "AIR" perspective and from a macOS perspective, it would need two sets of signing details - one before and one after the `-target bundle` bit.

But assuming you're doing it in two stages - firstly with IntelliJ using the signing details after the target, which is what AIR uses to create the signatures.xml file, and secondly using the separate `adt -sign` command with the certificate name from your Keychain store - then I would be wondering how you're checking for entitlements with the intermediate/unsigned app?

Not sure what you mean by the app changing, "inspect" in what sense? The app should still be a normal bundle structure and I think the Info.plist should still be text-based, but if not you can use xcode tools for it.

From what I remember, extra entitlements are requested via the application descriptor configuration per the link you gave, but these would then only be added into the executable file when it's signed. Unsigned binaries don't need entitlements (afaik?) but for distribution you would need to sign and notarise anyway?

There are a bunch of guides to this - the challenge being, things change over time particularly in Apple-world.. I think the latest guide was at https://github.com/airsdk/Adobe-Runtime-Support/discussions/2088#discussioncomment-11127165
But see other parts of https://github.com/airsdk/Adobe-Runtime-Support/discussions/2088 plus https://github.com/airsdk/Adobe-Runtime-Support/discussions/3387 and https://github.com/airsdk/Adobe-Runtime-Support/discussions/3075 perhaps?
